Description: Shut down the libev handlers on context destruction.
Origin: upstream; https://github.com/warmcat/libwebsockets/commit/d8383ca5fc9a868e70c7c307bf7f8f04a24b10ab
Author: gaby64 <gaby_64@live.ca>
Last-Update: 2015-12-25

--- a/lib/context.c
+++ b/lib/context.c
@@ -304,6 +304,11 @@
 			protocol++;
 		}
 	}
+#ifdef LWS_USE_LIBEV
+	ev_io_stop(context->io_loop, &context->w_accept.watcher);
+	if(context->use_ev_sigint)
+		ev_signal_stop(context->io_loop, &context->w_sigint.watcher);
+#endif /* LWS_USE_LIBEV */
 
 	lws_plat_context_early_destroy(context);
 	lws_ssl_context_destroy(context);
