Description: Recognize Debian GNU/kFreeBSD as FreeBSD-ish.
Forwarded: https://github.com/warmcat/libwebsockets/pull/383
Author: Peter Pentchev <roam@ringlet.net>
Last-Update: 2015-12-21

--- a/lib/lws-plat-unix.c
+++ b/lib/lws-plat-unix.c
@@ -201,7 +201,9 @@
 	int optval = 1;
 	socklen_t optlen = sizeof(optval);
 
-#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__) || \
+#if defined(__APPLE__) || \
+    defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || \
+    defined(__NetBSD__) || \
     defined(__OpenBSD__)
 	struct protoent *tcp_proto;
 #endif
@@ -213,7 +215,9 @@
 			       (const void *)&optval, optlen) < 0)
 			return 1;
 
-#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__) || \
+#if defined(__APPLE__) || \
+    defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || \
+    defined(__NetBSD__) || \
         defined(__CYGWIN__) || defined(__OpenBSD__)
 
 		/*
@@ -241,7 +245,9 @@
 
 	/* Disable Nagle */
 	optval = 1;
-#if !defined(__APPLE__) && !defined(__FreeBSD__) && !defined(__NetBSD__) && \
+#if !defined(__APPLE__) && \
+    !defined(__FreeBSD__) && !defined(__FreeBSD_kernel__) && \
+    !defined(__NetBSD__) && \
     !defined(__OpenBSD__)
 	if (setsockopt(fd, SOL_TCP, TCP_NODELAY, (const void *)&optval, optlen) < 0)
 		return 1;
