Description: Sort the source files for reproducible docs generation.
Bug-Debian: https://bugs.debian.org/808877
Forwarded: https://github.com/warmcat/libwebsockets/pull/401
Author: Peter Pentchev <roam@ringlet.net>
Last-Update: 2016-01-16

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -939,6 +939,7 @@
 	# TODO: Fix this on Windows.
 	message("Generating API documentation")
 	file(GLOB C_FILES "${PROJECT_SOURCE_DIR}/lib/*.c")
+	list(SORT C_FILES)
 	execute_process(COMMAND "${CMAKE_COMMAND}" -E make_directory "${PROJECT_BINARY_DIR}/doc/")
 
 	execute_process(
