From 2d69a41a790ab2c006cb46811beadd132286cbd5 Mon Sep 17 00:00:00 2001
From: Jarod Wilson <jarod@redhat.com>
Date: Thu, 11 Nov 2010 10:40:06 -0500
Subject: [PATCH 07/11] lirc_serial: fill in dev pointer

Signed-off-by: Jarod Wilson <jarod@redhat.com>
---
 drivers/lirc_serial/lirc_serial.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

--- a/drivers/lirc_serial/lirc_serial.c
+++ b/drivers/lirc_serial/lirc_serial.c
@@ -1289,6 +1289,7 @@ static int __init lirc_serial_init_modul
 	if (result < 0)
 		goto exit_serial_exit;
 	driver.features = hardware[type].features;
+	driver.dev = &lirc_serial_dev->dev;
 	driver.minor = lirc_register_driver(&driver);
 	if (driver.minor < 0) {
 		printk(KERN_ERR  LIRC_DRIVER_NAME
