From c9960a0f97bfee765a913212b2ca6cd26f22f1a5 Mon Sep 17 00:00:00 2001
From: Martijn de Gouw <martijn.de.gouw@prodrive.nl>
Date: Tue, 15 Jan 2013 20:13:49 +0100
Subject: [PATCH] Add support for 230400 baudrate

---
 src/modbus-rtu.c | 6 ++++++
 1 file changed, 6 insertions(+)

Index: libmodbus-3.0.6/src/modbus-rtu.c
===================================================================
--- libmodbus-3.0.6.orig/src/modbus-rtu.c
+++ libmodbus-3.0.6/src/modbus-rtu.c
@@ -394,6 +394,9 @@ static int _modbus_rtu_connect(modbus_t
     case 115200:
         dcb.BaudRate = CBR_115200;
         break;
+    case 230400:
+        dcb.BaudRate = 230400;
+        break;
     default:
         dcb.BaudRate = CBR_9600;
         printf("WARNING Unknown baud rate %d for %s (B9600 used)\n",
@@ -516,6 +519,9 @@ static int _modbus_rtu_connect(modbus_t
     case 115200:
         speed = B115200;
         break;
+    case 230400:
+        speed = B230400;
+        break;
     default:
         speed = B9600;
         if (ctx->debug) {
